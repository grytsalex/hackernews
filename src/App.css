div {
  width: 100%;
  height: 90vh;
  background: aquamarine;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}