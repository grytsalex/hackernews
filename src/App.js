// eslint-disable-next-line no-unused-vars
import React, { Component } from 'react';
import axios from 'axios';

import './App.css';
import ExplainBindingsComponent, { Search, Table, Button }  from './components';
import { DEFAULT_QUERY, DEFAULT_HPP, PATH_BASE, PARAM_SEARCH, PATH_SEARCH, PARAM_PAGE, PARAM_HPP } from './consts';


class App extends Component {
  _isMounted = false;

  constructor(props) {
    super(props)

      this.state = {
        results: null,
        searchKey: '',
        searchTerm: DEFAULT_QUERY,
        error: null,
      }

    this.needsToSearchTopStories = this.needsToSearchTopStories.bind(this);
    this.setSearchTopStories = this.setSearchTopStories.bind(this);
    this.getSearchTopStories = this.getSearchTopStories.bind(this);
    this.onSearchChange = this.onSearchChange.bind(this);
    this.onSearchSubmit = this.onSearchSubmit.bind(this);
    this.onDismiss = this.onDismiss.bind(this);
  };


  needsToSearchTopStories(searchTerm) {
      return !this.state.results[searchTerm]
  }

  setSearchTopStories(result) {
   const { hits, page } = result;
   const { searchKey, results } = this.state

    const oldHits = results && results[searchKey] ? results[searchKey].hits : [];

    const updatedHits = [
      ...oldHits,
      ...hits
    ];

    this.setState( { results: {...results, [searchKey]: { hits: updatedHits, page }}} )
  }

  onSearchSubmit(event) {
    const { searchTerm } = this.state;

    this.setState({ searchKey: searchTerm });

    if(this.needsToSearchTopStories(searchTerm)) {
      this.getSearchTopStories(searchTerm);
    }
    event.preventDefault()
  }

  onDismiss(id) {
    const { searchKey, results } = this.state;
    const { hits, page } = results[searchKey];

    const isNotId = item => item.objectID !== id
    const updatedHits = hits.filter(isNotId)

    this.setState({ results: { ...results, [searchKey]: { hits: updatedHits, page }}});
  }

  onSearchChange(event) {
    this.setState({ searchTerm: event.target.value });
  }

  getSearchTopStories(searchTerm, page = 0 ) {
    axios(`${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`)
    .then(result => this.setSearchTopStories(result.data))
    .catch(error => this._isMounted && this.setState({ error }))
  }

  componentDidMount() {

    this._isMounted = true;

    const { searchTerm } = this.state;
    this.setState({ searchKey: searchTerm })
    this.getSearchTopStories(searchTerm);
  }

  componentWillUnmount() {
    this._isMounted = false;
  }

  render() {

    const { searchTerm, results, searchKey, error } = this.state

    const page = (results && results[searchKey] && results[searchKey].page ) || 0
 
    const list = (results && results[searchKey] && results[searchKey].hits) || []


    return (
        <div className="page">
        <div className="interactions">
          <ExplainBindingsComponent/>
          <Search
            value={searchTerm}
            onChange={this.onSearchChange}
            onSubmit={this.onSearchSubmit}
          />
          </div>
          {error ? 
            <div className="interactions">
                <p>Something went wrong</p>
          </div> 
            :
          <Table
            list={list}
            onDismiss={this.onDismiss}
          />
          }
          <div className="interactions">
            <Button className="moreStories-btn" onClick={() => this.getSearchTopStories(searchTerm, page + 1)}>More stories</Button>
          </div>
        </div>
    );
  }
}

export default App