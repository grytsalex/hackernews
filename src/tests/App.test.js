import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import App from '../App';

describe('App', () => {
    it('render without error', () => {
        const div = document.createElement('div');
        ReactDOM.render(<App />, div);
        ReactDOM.unmountComponentAtNode(div);
    })

    test('should have snapshot', () => {
        const component = renderer.create(<App />)

        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    })
})