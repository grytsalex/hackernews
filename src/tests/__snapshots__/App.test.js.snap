// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should have snapshot 1`] = `
<div
  className="page"
>
  <div
    className="interactions"
  >
    <div
      className="bind"
    >
      <button
        onClick={[Function]}
        type="button"
      >
        Press here!
      </button>
    </div>
    <form
      onSubmit={[Function]}
    >
      <input
        onChange={[Function]}
        type="text"
        value="redux"
      />
      <button
        onClick={[Function]}
        type="submit"
      >
        Search
      </button>
    </form>
  </div>
  <div
    className="table"
  >
    <div
      className="table-header"
    >
      <span
        style={
          Object {
            "width": "40%",
          }
        }
      >
        <button
          className="button-inline"
          onClick={[Function]}
          style={
            Object {
              "alignItems": "center",
              "display": "flex",
              "justifyContent": "center",
            }
          }
          type="button"
        >
          <div
            style={
              Object {
                "paddingRight": "5px",
              }
            }
          >
            Title
          </div>
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-sort-down fa-w-10 fa-lg "
            data-icon="sort-down"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={
              Object {
                "paddingBottom": "5px",
              }
            }
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </button>
      </span>
      <span
        style={
          Object {
            "width": "30%",
          }
        }
      >
        <button
          className="button-inline"
          onClick={[Function]}
          style={
            Object {
              "alignItems": "center",
              "display": "flex",
              "justifyContent": "center",
            }
          }
          type="button"
        >
          <div
            style={
              Object {
                "paddingRight": "5px",
              }
            }
          >
            Author
          </div>
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-sort-down fa-w-10 fa-lg "
            data-icon="sort-down"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={
              Object {
                "paddingBottom": "5px",
              }
            }
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </button>
      </span>
      <span
        style={
          Object {
            "width": "10%",
          }
        }
      >
        <button
          className="button-inline"
          onClick={[Function]}
          style={
            Object {
              "alignItems": "center",
              "display": "flex",
              "justifyContent": "center",
            }
          }
          type="button"
        >
          <div
            style={
              Object {
                "paddingRight": "5px",
              }
            }
          >
            Comments
          </div>
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-sort-down fa-w-10 fa-lg "
            data-icon="sort-down"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={
              Object {
                "paddingBottom": "5px",
              }
            }
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </button>
      </span>
      <span
        style={
          Object {
            "width": "10%",
          }
        }
      >
        <button
          className="button-inline"
          onClick={[Function]}
          style={
            Object {
              "alignItems": "center",
              "display": "flex",
              "justifyContent": "center",
            }
          }
          type="button"
        >
          <div
            style={
              Object {
                "paddingRight": "5px",
              }
            }
          >
            Points
          </div>
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-sort-down fa-w-10 fa-lg "
            data-icon="sort-down"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={
              Object {
                "paddingBottom": "5px",
              }
            }
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </button>
      </span>
      <span
        style={
          Object {
            "display": "table",
            "margin": "auto",
            "width": "10%",
          }
        }
      >
        Archive
      </span>
    </div>
  </div>
  <div
    className="interactions"
  >
    <div
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "justifyContent": "center",
        }
      }
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-spinner fa-w-16 fa-lg "
        data-icon="spinner"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
      <p
        style={
          Object {
            "paddingLeft": "10px",
          }
        }
      >
        Loading...
      </p>
    </div>
  </div>
</div>
`;
